File: 959-3sum-with-multiplicity/3sum-with-multiplicity.java
class Solution {
    public int threeSumMulti(int[] nums, int target) {
        Arrays.sort(nums);
        int n = nums.length;long count=0;int i=0;
        while(i<n-2){
            int j=i+1;int k=n-1;
            int reqSum=target-nums[i];
            while(j<k){
                if(nums[j]+nums[k]==reqSum){
                    int jCount=1;int kCount=1;
                    if(nums[j]!=nums[k]){
                        while(nums[j]==nums[j+1] && j<k){
                            jCount++;
                            j++;
                        }
                        while(nums[k]==nums[k-1] && j<k){
                            kCount++;
                            k--;
                        }
                        count+=(long)(jCount*kCount);
                    }else{
                        int len = k-j+1;
                        count+=(long)((len*(len-1))/2);
                        break;
                    }
                    j++;//for continuation of loop.
                }else if(nums[j]+nums[k]<reqSum){
                    j++;
                }else{
                    k--;
                }
            }
            i++;
        }
        long mod = (long) 1e9+7;
        int ans = (int)(count%mod);
        return ans;
    }
}

File: 959-3sum-with-multiplicity/3sum-with-multiplicity.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumMulti = function(nums, target) {
    nums.sort((a, b) => a - b);
    const n = nums.length;
    let count = 0;
    const MOD = 1_000_000_007;

    for (let i = 0; i < n - 2; i++) {
        let j = i + 1;
        let k = n - 1;
        let reqSum = target - nums[i];

        while (j < k) {
            let currentSum = nums[j] + nums[k];
            if (currentSum === reqSum) {
                // Case 1: nums[j] and nums[k] are different
                if (nums[j] !== nums[k]) {
                    let jCount = 1;
                    let kCount = 1;
                    // Count occurrences of nums[j]
                    while (j + 1 < k && nums[j] === nums[j+1]) {
                        jCount++;
                        j++;
                    }
                    // Count occurrences of nums[k]
                    while (k - 1 > j && nums[k] === nums[k-1]) {
                        kCount++;
                        k--;
                    }
                    
                    count += (jCount * kCount);
                    j++; // Move j to find new distinct number
                    k--; // Move k to find new distinct number
                } 
                // Case 2: nums[j] and nums[k] are the same
                else { // nums[j] === nums[k]
                    // All elements from j to k are identical
                    const length = k - j + 1;
                    // Choose 2 elements from 'length' identical elements
                    count += (length * (length - 1)) / 2;
                    break; // All valid pairs for current i are found
                }
            } else if (currentSum < reqSum) {
                j++;
            } else { // currentSum > reqSum
                k--;
            }
        }
    }
    
    return count % MOD;
};